gcovr/coverage.py

```diff
@@ -314,6 +318,110 @@ def call_coverage(self) -> CoverageStat:
+                subdirs[key] = DirectoryCoverage.new_empty()
+                subdirs[key].filename = key
+                subdirs[key].parent_key = DirectoryCoverage.directory_key(
+                    key, root_filter
+                )
+
+            if newchild:
+                if newchild not in subdirs[key].children:
+                    subdirs[key].children.append(newchild)
+            else:
+                filecov.parent_key = key
+                subdirs[key].children.append(filecov)
+
+            subdirs[key].stats += SummarizedStats.from_file(filecov)
+            DirectoryCoverage.add_directory_coverage(
+                subdirs, key, filecov, root_filter, subdirs[key]
+            )
+
+    @staticmethod
+    def collapse_subdirectories(
+        subdirs: CovData_subdirectories, root_filter: re.Pattern
+    ) -> None:
+        collapse_dirs = set()
+        root_key = DirectoryCoverage.directory_root(subdirs, root_filter)
+        for key, value in subdirs.items():
+            if (
+                isinstance(value, DirectoryCoverage)
+                and len(value.children) == 1
+                and not key == root_key
+            ):
+                while True:
+                    parent_key = DirectoryCoverage.directory_key(key, root_filter)
+                    if parent_key not in collapse_dirs or parent_key == root_key:
+                        break
+                if parent_key:
+                    newchildren = [
+                        child
+                        for child in subdirs[parent_key].children
+                        if child != value
+                    ]
+                    orphan = value.children[0]
+                    orphan.parent_key = parent_key
+                    newchildren.append(orphan)
+                    subdirs[parent_key].children = newchildren
+
+                    collapse_dirs.add(key)
+        for key in collapse_dirs:
+            del subdirs[key]
+
+    @staticmethod
+    def from_covdata(
+        covdata: CovData, sorted_keys: Iterable, root_filter: re.Pattern
+    ) -> CovData_subdirectories:
+        subdirs = OrderedDict()
+        for key in sorted_keys:
+            filecov = covdata[key]
+            DirectoryCoverage.add_directory_coverage(
+                subdirs, filecov.filename, filecov, root_filter
+            )
+        return subdirs
+
+    @staticmethod
+    def directory_key(filename: str, root_filter: re.Pattern):
+        key = os.path.dirname(filename.replace("\\", "/"))
+        if root_filter.search(key + "/") and key != filename:
+            return key
+        return None
+
+    @staticmethod
+    def directory_root(subdirs: CovData_subdirectories, root_filter: re.Pattern) -> str:
+        key = next(iter(subdirs))
+        while True:
+            next_key = DirectoryCoverage.directory_key(key, root_filter)
+            if not next_key:
+                return key
+            else:
+                key = next_key
```

Comment:
The error on Windows is here but I need to investigate deeper what you're doing here.

Reply:
The iterator is at the end and the exception isn't caught.